### PR TITLE
Fix mapping of timepipelined input routes (O2-1180)

### PR DESCRIPTION
**Pull request by knopers8**
@ktf 
There is a problem if one specifies a device with a timepipeline input and another which is not timepipelined (e.g. a timer). If I check if the timer is valid, `InputRecord::getPos` returns a wrong position, which results later in the following error:
```
[28718:dataReceiver]: [10:38:52][ERROR] Exception caught: Unknown argument requested time - Unknown message requested at position 2
```
That fix is a bit dirty and I am not sure if it doesn't break anything else. Let's see what CI says.

### CODE REVIEW

File: Framework/Utils/test/test_RootTreeWriter.cxx [hunk 1]
```diff
@@ -188,14 +188,14 @@ BOOST_AUTO_TEST_CASE(test_RootTreeWriter)
   // temporary argument is still in memory
   // FIXME: check why the compiler does not detect this
   std::vector<InputRoute> schema = {
-    {InputSpec{"input1", "TST", "INT"}, 0, "input1", 0},       //
-    {InputSpec{"input2", "TST", "CONTAINER"}, 1, "input2", 0}, //
-    {InputSpec{"input3", "TST", "CONTAINER"}, 2, "input3", 1}, //
```
Comment (knopers8) [inline Framework/Utils/test/test_RootTreeWriter.cxx]: @ktf Was that actually a correct usage of `timeslice` before? As far as I understand (perhaps I am wrong), time-pipeline inputs can have multiple InputRoutes, but can correspond to one and only InputSpec. So in a real DPL device, the user algorithm would see a message coming from only one of the two InputRoutes at a time. Is that correct?

**Comment by knopers8** [comment]
@matthiasrichter Could you have a look at this? It seems that this might interfere with https://github.com/AliceO2Group/AliceO2/pull/2886

**Comment by ktf** [comment]
As discussed privately, I agree that the semantic of "getPos()" should be "binding to inputSpec space" rather than "binding to InputRoute Space". @matthiasrichter we think the test was relying on a wrong assumption. Can you cross check we are not missing anything?

**Comment by knopers8** [comment]
With the new version of the fix, testDataAllocator and testStaggeringWorkflow are failing, which might be related, I will have a look.

**Comment by knopers8** [comment]
It seems that in test_DataAllocator the InputsSchema (so InputRoute space) has the inputs in the reverse order as in the InputSpecs space:
With this PR - position is route.inputSpecIndex:
```
[11563:sink]: [11:45:36][INFO] input1 getPos: 0
[11563:sink]: [11:45:36][INFO] input2 getPos: 1
[11563:sink]: [11:45:36][INFO] input3 getPos: 2
```
Without - position is index in InputsSchema:
```
[12450:sink]: [11:50:16][INFO] input1 getPos: 15
[12450:sink]: [11:50:16][INFO] input2 getPos: 14
[12450:sink]: [11:50:16][INFO] input3 getPos: 13
```
Not sure why yet, I will come back to it on Monday

**Comment by knopers8** [comment]
@ktf `inputSpecIndex` in fact doesn't always map to the right position in `InputRecord::mSpan`. It comes from the edge matching phase and these input edges are sorted in `WorkflowHelpers::sortEdges` (not sure what for). On the other hand, the `InputMatchers` and `InputRoutes` follow the original, unsorted order.
It seems that usually the sorted input edges have the same order as `InputRoutes` (this is why it worked when we were debugging `test_RootTreeWriter`), but it is not always the case. With that, we should either make sure that `inputRoutes` follow the order the same order as the sorted input edges, or use the original, blind version of the fix:
```
int InputRecord::getPos(const char* binding) const
{
  size_t spanID = -1;
  for (size_t i = 0; i < mInputsSchema.size(); ++i) {
    auto& route = mInputsSchema[i];

    if (route.timeslice != 0) {
      continue;
    }
    spanID++;
    if (route.matcher.binding == binding) {
      return spanID++;
    }
  }
  return -1;
}
```

**Comment by matthiasrichter** [comment]
I think we have not been using time pipelining so far in a real example, I have planned to use it in the TPC reco workflow for a long time. It's good o get this working I will have a look whats happening in the unit tests.

**Comment by knopers8** [comment]
@matthiasrichter Hello, did you get a chance to have a look at it?

**Comment by knopers8** [comment]
This PR is still relevant, we wait for @matthiasrichter to have a look on the failing unit tests.

**Comment by ktf** [comment]
This is still needed and we need to understand how to integrate it.

**Comment by matthiasrichter** [comment]
I only now took notice of this. From a first look it seems that the failing unit tests indicate real errors, e.g. in `test_DataAllocator ` a message can not be retrieved on the receiver side
```
[15647:sink]: [11:17:45][INFO] extracting o2::test::TriviallyCopyable from input1
[15647:sink]: [11:17:45][ERROR] Exception caught: Can not extract a plain object from serialized message
```
Not sure how this can happen.

**Pull request closed without merging**